drivers/serial/uart_ifx_cat1.c

```diff
@@ -0,0 +1,506 @@
+
+	cyhal_uart_enable_event(&data->obj,
+				(cyhal_uart_event_t) CYHAL_UART_IRQ_RX_NOT_EMPTY,
+				IFX_CAT1_UART_INTERRUPT_PRIORITY, 1);
+}
+
+/* Disable TX interrupt */
+static void ifx_cat1_uart_irq_rx_disable(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+
+	cyhal_uart_enable_event(&data->obj,
+				(cyhal_uart_event_t) CYHAL_UART_IRQ_RX_NOT_EMPTY,
+				IFX_CAT1_UART_INTERRUPT_PRIORITY, 0);
+}
+
+/* Check if UART RX buffer has a received char */
+static int ifx_cat1_uart_irq_rx_ready(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+
+	return cyhal_uart_readable(&data->obj) ? 1 : 0;
+}
+
+/* Enable Error interrupts */
+static void ifx_cat1_uart_irq_err_enable(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+
+	cyhal_uart_enable_event(&data->obj, (cyhal_uart_event_t)
+				(CYHAL_UART_IRQ_TX_ERROR | CYHAL_UART_IRQ_RX_ERROR),
+				IFX_CAT1_UART_INTERRUPT_PRIORITY, 1);
+}
+
+/* Disable Error interrupts */
+static void ifx_cat1_uart_irq_err_disable(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+
+	cyhal_uart_enable_event(&data->obj, (cyhal_uart_event_t)
+				(CYHAL_UART_IRQ_TX_ERROR | CYHAL_UART_IRQ_RX_ERROR),
+				IFX_CAT1_UART_INTERRUPT_PRIORITY, 0);
+}
+
+/* Check if any IRQs is pending */
+static int ifx_cat1_uart_irq_is_pending(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+	uint32_t intcause = Cy_SCB_GetInterruptCause(data->obj.base);
+
+	return (int) (intcause & (CY_SCB_TX_INTR | CY_SCB_RX_INTR));
+}
+
+/* Start processing interrupts in ISR.
+ * This function should be called the first thing in the ISR. Calling
+ * uart_irq_rx_ready(), uart_irq_tx_ready(), uart_irq_tx_complete()
+ * allowed only after this.
+ */
+static int ifx_cat1_uart_irq_update(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+	int status = 1;
+
+	if (((ifx_cat1_uart_irq_is_pending(dev) & CY_SCB_RX_INTR) != 0u) &&
+	    (Cy_SCB_UART_GetNumInRxFifo(data->obj.base) == 0u)) {
+		status = 0;
+	}
+
+	return status;
+}
+
+static void ifx_cat1_uart_irq_callback_set(const struct device *dev,
+					   uart_irq_callback_user_data_t cb,
+					   void *cb_data)
+{
+	struct ifx_cat1_uart_data *data = dev->data;
+	cyhal_uart_t *uart_obj = &data->obj;
+
+	/* Store user callback info */
+	data->irq_cb = cb;
+	data->irq_cb_data = cb_data;
+
+	/* Register a uart general callback handler  */
+	cyhal_uart_register_callback(uart_obj, _uart_event_callback_irq_mode, (void *) dev);
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+
+static int ifx_cat1_uart_init(const struct device *dev)
+{
+	struct ifx_cat1_uart_data *const data = dev->data;
+	const struct ifx_cat1_uart_config *const config = dev->config;
+	cy_rslt_t result;
+	int ret;
+
+	/* Default Counter configuration structure */
+	static const cy_stc_scb_uart_config_t _cyhal_uart_default_config = {
```

Comment:
static on a multi-instance driver?

Reply:
this is `static const` to be sure that this structure locates in RO section..

Reply:
Personally I like to have `static` (and `static const`) outside of function body. It is easier to locate those differentiate from stack defined variables. It also clearly makes such structure a "configuration struct" and outside of function body, which is usually a "logic implementation".

Not sure whether we have guidelines in Zephyr that say the same.